File: src/pipeline/kalman.py
"""Filtre de Kalman pour le suivi du quadrilatère d'une plaque (4 points)."""
from __future__ import annotations

import numpy as np
from filterpy.kalman import KalmanFilter


def make_kalman_quad(polygon) -> KalmanFilter:
    """Crée un filtre de Kalman à vitesse constante pour un quadrilatère.

    État   : [x1,y1, x2,y2, x3,y3, x4,y4,  vx1,vy1, vx2,vy2, vx3,vy3, vx4,vy4]
    Mesure : [x1,y1, x2,y2, x3,y3, x4,y4]
    """
    kf = KalmanFilter(dim_x=16, dim_z=8)

    # Matrice de transition F : position += vitesse
    kf.F = np.eye(16)
    for i in range(8):
        kf.F[i, i + 8] = 1.0

    # Matrice d'observation H : on mesure les 8 coordonnées de position
    kf.H = np.zeros((8, 16))
    for i in range(8):
        kf.H[i, i] = 1.0

    # Bruits
    kf.R *= 2       # bruit de mesure (confiance accordée à YOLO)
    kf.P *= 1       # incertitude initiale
    kf.Q *= 0.01    # bruit de processus (mouvement supposé régulier)

    # Position initiale
    pts = np.array(polygon, dtype=np.float32).flatten()
    kf.x[:8] = pts.reshape(8, 1)

    return kf


def kalman_predict_quad(kf: KalmanFilter) -> list:
    """Retourne le polygone prédit [[x1,y1], [x2,y2], [x3,y3], [x4,y4]]."""
    kf.predict()
    return kf.x[:8].flatten().reshape(4, 2).astype(int).tolist()


def kalman_update_quad(kf: KalmanFilter, polygon) -> list:
    """Met à jour le filtre avec le polygone mesuré — retourne le polygone lissé."""
    pts = np.array(polygon, dtype=np.float32).flatten().reshape(8, 1)
    kf.update(pts)
    return kf.x[:8].flatten().reshape(4, 2).astype(int).tolist()

File: cpp/src/pipeline/kalman.cpp
#include "pipeline/kalman.hpp"

namespace lapi {

namespace {

// Troncature vers zéro, comme ndarray.astype(int).
Quad state_to_quad(const cv::Mat& x) {
    Quad q;
    for (int i = 0; i < 4; ++i) {
        q[i].x = static_cast<float>(static_cast<int>(x.at<double>(2 * i, 0)));
        q[i].y = static_cast<float>(static_cast<int>(x.at<double>(2 * i + 1, 0)));
    }
    return q;
}

cv::Mat quad_to_measurement(const Quad& polygon) {
    cv::Mat z(8, 1, CV_64F);
    for (int i = 0; i < 4; ++i) {
        // float32 puis double, comme np.array(polygon, dtype=np.float32)
        z.at<double>(2 * i, 0) = static_cast<double>(polygon[i].x);
        z.at<double>(2 * i + 1, 0) = static_cast<double>(polygon[i].y);
    }
    return z;
}

}  // namespace

KalmanQuad make_kalman_quad(const Quad& polygon) {
    KalmanQuad kf;
    kf.F = cv::Mat::eye(16, 16, CV_64F);
    for (int i = 0; i < 8; ++i)
        kf.F.at<double>(i, i + 8) = 1.0;

    kf.H = cv::Mat::zeros(8, 16, CV_64F);
    for (int i = 0; i < 8; ++i)
        kf.H.at<double>(i, i) = 1.0;

    kf.R = cv::Mat::eye(8, 8, CV_64F) * 2.0;     // confiance accordée à YOLO
    kf.P = cv::Mat::eye(16, 16, CV_64F);         // incertitude initiale
    kf.Q = cv::Mat::eye(16, 16, CV_64F) * 0.01;  // mouvement supposé régulier

    kf.x = cv::Mat::zeros(16, 1, CV_64F);
    quad_to_measurement(polygon).copyTo(kf.x.rowRange(0, 8));
    return kf;
}

void kalman_predict(KalmanQuad& kf) {
    kf.x = kf.F * kf.x;
    kf.P = kf.F * kf.P * kf.F.t() + kf.Q;
}

Quad kalman_predict_quad(KalmanQuad& kf) {
    kalman_predict(kf);
    return state_to_quad(kf.x);
}

Quad kalman_update_quad(KalmanQuad& kf, const Quad& polygon) {
    // Équations de filterpy, forme de Joseph pour P.
    const cv::Mat z = quad_to_measurement(polygon);
    const cv::Mat y = z - kf.H * kf.x;
    const cv::Mat PHT = kf.P * kf.H.t();
    const cv::Mat S = kf.H * PHT + kf.R;
    const cv::Mat K = PHT * S.inv(cv::DECOMP_LU);
    kf.x = kf.x + K * y;
    const cv::Mat I_KH = cv::Mat::eye(16, 16, CV_64F) - K * kf.H;
    kf.P = I_KH * kf.P * I_KH.t() + K * kf.R * K.t();
    return state_to_quad(kf.x);
}

}  // namespace lapi
